_SPIKE_.md

```diff
@@ -0,0 +1,496 @@
+  ): Promise<void>;
+  // ^^ nothing is returned, the source models are updated in-place
+}
+```
+
+With a bit of luck, we will be able to use these resolvers for GraphQL too.
+However, such use-case is out of scope of this spike.
+
+### 2. Base repository classes handle inclusions via resolvers
+
+Each repository class (e.g. `DefaultCrudRepository` from our legacy juggler
+bridge) should maintain a map containing inclusion resolvers for each relation
+that is allowed to be included.
+
+Conceptually:
+
+```ts
+export class DefaultCrudRepository {
+  public inclusionResolvers: {[key: string]: InclusionResolver};
+
+  // ...
+}
+```
+
+**IMPORTANT:** To support use cases where no custom repository class is created
+and applications/extensions are instantiating `DefaultCrudRepository` class
+directly, it's important to expose `inclusionResolvers` as a public property.
+
+When a repository method like `find`, `findOne` and `findById` is called to
+query the database, it should use use `inclusionResolvers` map to fetch any
+related models requested by `filter.include`.
+
+Conceptually:
+
+```ts
+export class DefaultCrudRepository<T, Relations> {
+  // ...
+
+  async find(
+    filter?: Filter<T>,
+    options?: Options,
+  ): Promise<(T & Relations)[]> {
+    const models = await ensurePromise(
+      this.modelClass.find(this.normalizeFilter(filter), options),
+    );
+    const entities = this.toEntities(models);
+    // NEW LINE ADDED TO RESOLVE INCLUDED MODELS:
+    return this.includeRelatedModels(entities, filter, options);
+  }
+
+  protected async includeRelatedModels(
+    entities: T[],
+    filter?: Filter<T>,
+    options?: Options,
+  ): Promise<(T & Relations)[]> {
+    const result = entities as (T & Relations)[];
+
+    // process relations in parallel
+    const resolveTasks = filter.include.map(i => {
+      const relationName = i.relation;
+      const handler = this.inclusionResolvers[relationName];
+      return handler.fetchIncludedModels(entities, i, options);
+    });
+    await Promise.all(resolveTasks);
+
+    return result;
+  }
+}
+```
+
+### 3. Model-specific repositories register inclusion resolvers
+
+Model-specific repository classes (e.g. `CategoryRepository`) should register
+inclusion resolvers for model relations, similarly to how we are creating
+relation-repository factories now.
+
+Conceptually:
+
+```ts
+export class CategoryRepository extends DefaultCrudRepository {
+  products: HasManyRepositoryFactory<Product, typeof Category.prototype.id>;
+
+  constructor(
+    dataSource: juggler.DataSource,
+    productRepositoryGetter: Getter<ProductRepository>,
+  ) {
+    super(Category, dataSource);
+
+    // we already have this line to create HasManyRepository factory
+    this.products = this.createHasManyRepositoryFactoryFor(
+      'todos',
+      productRepositoryGetter,
+    );
+
+    // add this line to register inclusion resolver
+    this.registerHasManyInclusion('products', this.productRepositoryGetter);
```

Comment:
As we already set up `this.products` as a factory to create a `HasManyRepository`, can we directly use it for the `inclusionResolver`?

Reply:
That's an interesting idea! IIUC, you are suggesting that `this.products` (`HasManyRepositoryFactory`) can implement also `InclusionResolver` API, therefore this new line would become:

```ts
this.registerInclusion(this.products);
```

I like that approach! :+1:

Reply:
In https://github.com/strongloop/loopback-next/pull/3387#discussion_r305021466, you proposed to change resolvers from an interface/class to a simple function. To accommodate that, I think we will have to register resolvers slightly differently, e.g.

```ts
this.registerInclusion('products', this.products.inclusionResolver);
```

Reply:
Done in e1f6745a8